README.md
## Such Wow

[![License: Unlicense](https://img.shields.io/badge/license-Unlicense-blue.svg)](http://unlicense.org/)
[![Current Version](https://img.shields.io/clojars/v/marick/suchwow.svg)](https://clojars.org/marick/suchwow)

The [API documentation index](http://marick.github.io/suchwow) gives an overview of what this library offers.

Snippets from serious work, wrapped in a whimsical container. A bit of a tribute to the spirit of [_why](http://en.wikipedia.org/wiki/Why_the_lucky_stiff), but with a [Shiba Inu](http://en.wikipedia.org/wiki/Shiba_Inu) instead of [foxes](http://mislav.uniqpath.com/poignant-guide/images/the.foxes-3.png).

This package offers three types of functions:
* `clojure.core` functions, but with better documentation (including examples).
* Variants of `clojure.core` functions that accept more kinds of inputs.
* A grab-bag of useful functions that, importantly, you can copy into your own code without worrying about licenses or giving credit or any of that. As a programmer trying to get work done, I use this library and others to create a ["favorite functions" namespace](https://github.com/marick/clojure-commons/blob/master/src/commons/clojure/core.clj) that I `use` everywhere.

![By Euterpia (Own work, CC0), via Wikimedia Commons](http://upload.wikimedia.org/wikipedia/commons/thumb/d/df/Doge_homemade_meme.jpg/256px-Doge_homemade_meme.jpg)     
[via Euterpia](http://commons.wikimedia.org/wiki/File:Doge_homemade_meme.jpg)

[![Build Status](https://travis-ci.org/marick/suchwow.png?branch=master)](https://travis-ci.org/marick/suchwow)

## Such Usage

Available via [clojars](https://clojars.org/marick/suchwow) for Clojure 1.7+  
For lein: [marick/suchwow "6.0.0"]

[Much API doc](http://marick.github.io/suchwow/)

Copy the source if you want, do the normal `(:require
[such.types :as wow])` thing, or create your own `commons.clojure.core`
namespace with all the things you think should be packaged with Clojure. 

The files [test/such/clojure/f_immigration.clj](https://github.com/marick/suchwow/blob/master/test/such/f_immigration.clj) and [commons.clojure.core](https://github.com/marick/clojure-commons/blob/master/src/commons/clojure/core.clj) show how to arrange for that last.


## Such License

This software is covered by the [Unlicense](http://unlicense.org/)
and, as such, is in the public domain.

## Such Contributors

* Alex Miller
* Bahadir Cambel
* Børge Svingen
* Brian Marick

## Such Contributing

Pull requests accepted, provided:

1. Your contribution has tests. In keeping with the spirit of the library, they
   don't even have to be written with
   [Midje](https://github.com/marick/Midje), since Midje can run
   clojure.test tests.

2. Your contribution doesn't depend on anything other than Clojure itself.

3. You have the right to put your contribution into the public domain.

    To allow me to be a teensy bit scrupulous, please include the following text in
    the comment of your pull request:

    > I dedicate any and all copyright interest in this software to the
    > public domain. I make this dedication for the benefit of the public at
    > large and to the detriment of my heirs and successors. I intend this
    > dedication to be an overt act of relinquishment in perpetuity of all
    > present and future rights to this software under copyright law.



CHANGELOG.md
# Change Log
This project adheres to [Semantic Versioning](http://semver.org/).
See [here](http://keepachangelog.com/) for the change log format.

## [6.0.3] - 2021-03-17
- bump libs

## [6.0.2] - 2017-12-27
- Move `defalias` and `alias-var` code over from `clojure-commons`

## [6.0.1] - 2017-12-27
- Upgrade to Specter 1.0.4 to avoid `any? in com.rpl.specter.impl` warning
- Fix reflection warnings
- Bump libs

## [6.0.0] - 2016-10-26
- Upgrade to Specter 0.13
- ... which caused Clojure 1.6 to be abandoned, so major version bump.

## [5.2.4]
- Update dependencies.

## [5.2.3]
- Fixes bad `ns` form. @puredanger

## [5.2.1]
- Specter 0.12.0 broke downstream dependencies. Reverting to 0.11.2

## [5.2.0]
- ADD: Better error message when trying to immigrate a variable that does not exist.
- Update dependencies

## [5.1.4]
- CHANGE: Guard against Clojure 1.9's `any?` - Børge Svingen

## [5.1.3]
- CHANGE: Bump structural typing dependency

## [5.1.2]
- CHANGE: Update `structural-typing` because of annoying circular dependency.

## [5.1.1]
- CHANGE: Update versions, most notably `specter`, but also `environ`,
  `combinatorics`, `structural-typing`, and `compojure`.

## [5.1]
- ADD: such.imperfection has variants that print to *err* of these functions:
  `pr`, `prn`, `print`, `println`, `pprint`, `-pprint-`, `-prn-`, `tag-`, and `-tag`.

## [5.0]
- CHANGED: No longer support Clojure 1.5
- ADD: such.control-flow/let-maybe
- ADD: An *experimental* such.relational namespace. See the API docs and [the wiki](https://github.com/marick/suchwow/wiki/such.relational)

## [4.4.3]
- DEPRECATION: In upcoming release, immigration functions will not `require` namespaces themselves.

## [4.4.2]

- bump versions of dependencies, including commons-codec

## [4.4.1]

- Use commons-codec 1.6 instead of 1.10. ring-codec uses 1.6. Since it's more popular,
  we might as well just track its preference to avoid shoving version conflicts into
  people's faces.

## [4.4.0]

### Added

- such.maps now supplies Clojure 1.7's `update` under Clojure 1.6.
- -tag and tag- can now take any value, not just a string.

## [4.3.0]

### Added

- Functions in such.imperfection and such.readable can now be used with Clojure 1.6.

## [4.2.0]

### Added
- `such.metadata/contains?`

### Fixed
- All the `such.immigration` functions now do requires at compile time.
- Workaround to prevent Codox from crashing on vars imported with `import-prefixed-vars`.

## [4.1.0] - 2015-08-19

### Added
- `imperfection` namespace for debugging and testing functions.

## [4.0.1] - 2015-08-14

### Fixed

- The generated `api` function didn't have `:arglist` metadata.

## [4.0.0] - 2015-08-14

### Changed

- Names in previous were bad: such.api -> such.doc, `open` -> `api`.

## [3.5.0] - 2015-08-14

### Added
- docs for `reduce`, `reductions`, `reduce-kv`, and `map-indexed`.
- `such.api`

## [3.4.0] - 2015-07-31

### Added
- `such.metadata` namespace

## [3.3.0] - 2015-07-11

### Added
- Docs for `sequential?`, `cond->`, and `cond->>`
- `pred:not-any?`, `pred:none-of?`

### Deprecated
- `any-pred` is now `pred:any?`
- `wrap-pred-with-catcher` is now `pred:exception->false`




project.clj
(defproject marick/suchwow "6.0.3"
  :description "Such functions! Such doc strings! Much utility!"
  :url "https://github.com/marick/suchwow"
  :pedantic? :warn
  :license {:name "The Unlicense"
            :url "http://unlicense.org/"
            :distribution :repo}

  :dependencies [[org.clojure/clojure "1.10.3"]
                 [potemkin "0.4.5" :exclusions [org.clojure/clojure]]
                 [com.rpl/specter "1.1.3" :exclusions [org.clojure/clojure org.clojure/clojurescript]]
                 [environ "1.2.0" :exclusions [org.clojure/clojure]]
                 [commons-codec/commons-codec "1.15"]]

  :repl-options {:init (do (require 'such.doc)
                           (such.doc/apis))}

  :profiles {:dev {:dependencies [[midje "1.9.10" :exclusions [org.clojure/clojure]]
                                  [org.clojure/math.combinatorics "0.1.6"]
                                  [org.clojure/data.json "1.1.0"]
                                  ;; Including compojure so that `lein ancient` will
                                  ;; tell us to upgrade, which might alert us that
                                  ;; compojure now depends on a more-modern version of
                                  ;; commons-codec.
                                  [marick/structural-typing "2.0.5" :exclusions [marick/suchwow]]
                                  [compojure "1.6.2" :exclusions [org.clojure/clojure]]]}
             :1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
             :1.8 {:dependencies [[org.clojure/clojure "1.8.0"]]}
             :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}}

  :plugins [[lein-midje "3.2.1"]
            [codox "0.8.11"]]

  :codox {:src-dir-uri "https://github.com/marick/suchwow/blob/master/"
          :src-linenum-anchor-prefix "L"
          :output-dir "/var/tmp/suchwow-doc"
          :defaults {:doc/format :markdown}}

  :aliases {"compatibility" ["with-profile" "+1.7:+1.8:+1.9" "midje" ":config" ".compatibility-test-config"]
            "travis" ["with-profile" "+1.7:+1.8:+1.9" "midje"]}

  ;; For Clojure snapshots
  :repositories {"sonatype-oss-public" "https://oss.sonatype.org/content/groups/public/"}
  :deploy-repositories [["releases" :clojars]])
